PScoreAssocUtil
===============

A java based helper utility for SCOREASSOC that generates simulated case control studies which can be used to evaluate the effects of various parameters on the statistical power of SCOREASSOC's ability to detect causal variants.




This is a command line utility that requires the following parameters:


1. the input VCF with a region as must be specified as:
*inputvcf yourVCF.vcf.gz region 16 12345 12346


2a. the number of case/control individuals to be simulated:
*simulate 100 100


2b. the number of simulated studies to be generated
*numsims 5


3. a variants file that has effect sizes and optionally frequencies defined for all variants that will confer an effect on the phenotype:
*variants PSCORE_VAR.csv


4. specify temporary simulation output files for SCOREASSOC to be able read in for its analysis:
*simoutput myOutput.dat myOutput.phe 


5. variant consequences file generated from Variants effect predictor generated with the following settings:
- http://grch37.ensembl.org/Homo_sapiens/Tools/VEP
- RefSeq transcripts
- add: Filters / Restrict results: show most severe consequence (otherwise it will have at least 2 entry for each
- upload VCF, then save it as .txt file
*vepinput VEP_NODRARE_COMBINED.csv


6. PSCOREASSOC location
 *pscorreassoc pscoreassoc.exe


7. the temporary weights file to be used by PSCOREASSOC:
 *weightsoutput nod2.annot nod2weights.txt 


8. the input Variant Frequency weight factor to be used by PSCOREASSOC
*weightfactor 10


9. the input Variant consequence weight map to be used by PSCOREASSOC
Note:  To produce this, either write it manually, or you could use the output generated by '*weightsoutput's second parameter (nod2weights.txt above). 
This will be a template file with all the variant consequence types present in the dataset with a default weight of 1, that you could manually edit to your purposes.
*weightmap weigthmap.csv 


10. Where the signed logP values are to be saved (coming from PSCOREASSOC)
*pvaloutput logPValues.txt


11. If and R-script is to be produced that is attempted to be executed (R 3.1.1 must be installed)
 *routput 


- optional: specify random seed for reproducible results:
*seed 1





EXAMPLE:

java -jar PScoreAssocUtil.jar *inputvcf g:/!!!0_EclipseWorkspace/ALL.chr16.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz *simulate 100 100 *region 16 50731049 50766987 *variants PSCORE_VAR.csv *seed 1 *simoutput myOutput.dat myOutput.phe *vepinput VEP_NODRARE_COMBINED.csv *weightsoutput nod2.annot nod2weights.txt *numsims 5 *pscorreassoc pscoreassoc.exe *pvaloutput logPValues.txt *routput *weightmap weigthmap.csv *weightfactor 10


- the above will load in a VCF (chrom 16 from the 1000gp), extract variants between positions 50731049-50766987, select the effect-variants specified by 'PSCORE_VAR.csv', load in variant consequences from file 'VEP_NODRARE_COMBINED.csv', 5 simualted studies are generated with 100 cases and 1000 controls, output simulation files ('myOutput.dat' and 'myOutput.phe') for SCOREASSOC, which is then called with a frequency weightfactor of 10 with the temporary weight file 'nod2.annot'. The analysis' signed logP values are then saved into 'logPValues.txt', after which R is called to produce a plot saved as a .png.
(random seed at 1)



PS:
an example of how the application could be batch run from a cluster can be found under:

EXAMPLE/batchRare.sh